build_runner/lib/src/build/asset_graph/nodes.dart

```diff
@@ -0,0 +1,239 @@
+      _sortedFileIdsByPackage = null;
+    }
+
+    return updatedNode;
+  }
+
+  /// Updates a node in the graph with [updates].
+  ///
+  /// If it does not exist, does nothing and returns `null`.
+  ///
+  /// If it does exist, returns the updated node.
+  AssetNode? updateNodeIfPresent(
+    AssetId id,
+    void Function(AssetNodeBuilder) updates,
+  ) {
+    final node = get(id);
+    if (node == null) return null;
+    final updatedNode = node.rebuild(updates);
+    _nodes[id] = updatedNode;
+
+    if (node.inputs != updatedNode.inputs) {
+      _outputs = null;
+    }
+    if (node.isFile != updatedNode.isFile) {
+      _sortedFileIdsByPackage = null;
+    }
+
+    return updatedNode;
+  }
+
+  /// Adds [node] to the graph if it doesn't exist.
+  ///
+  /// Throws a [StateError] if it already exists in the graph.
+  ///
+  /// Returns the updated node: if it replaces an [AssetNode.missingSource] then
+  /// `outputs` and `primaryOutputs` are copied to it from that.
+  AssetNode add(AssetNode node) {
+    if (node.isFile) _sortedFileIdsByPackage = null;
+    final existing = get(node.id);
+    if (existing != null) {
+      if (existing.type == NodeType.missingSource) {
+        _nodes.remove(existing.id);
+        node = node.rebuild((b) {
+          b.primaryOutputs.addAll(existing.primaryOutputs);
+        });
+      } else {
+        throw StateError(
+          'Tried to add node ${node.id} to the asset graph but it already '
+          'exists.',
+        );
+      }
+    }
+    _nodes[node.id] = node;
+    if (node.inputs?.isNotEmpty ?? false) {
+      _outputs = null;
+    }
+
+    return node;
+  }
+
+  /// Removes [id] from the graph, if present.
+  void remove(AssetId id) {
+    final removed = _nodes.remove(id);
+    if (removed != null && removed.isFile) {
+      _sortedFileIdsByPackage = null;
+    }
+  }
+
+  /// All nodes in the graph.
+  Iterable<AssetNode> get allNodes => _nodes.values;
+
+  /// All IDs in `package` that refer to files and match.
+  ///
+  /// A node refers to a file if [AssetNode.isFile].
+  ///
+  /// If [glob] is passed, return only IDs for which the glob matches the path.
+  Iterable<AssetId> packageFileIds(String package, {Glob? glob}) {
+    _sortedFileIdsByPackage ??= _computeSortedFileIdsByPackage();
+    final list = _sortedFileIdsByPackage![package];
+    if (list == null) return const [];
+    if (glob == null) return list;
+
+    final simplePrefix = _simpleGlobPrefix(glob);
+    Iterable<AssetId> result = list;
+    if (simplePrefix.isNotEmpty) {
+      result = list.filterToPrefix(simplePrefix);
+    }
+
+    return result.where((id) => glob.matches(id.path));
+  }
+
+  /// The prefix of the [glob] pattern up to but not including the first special
+  /// character.
+  String _simpleGlobPrefix(Glob glob) {
+    final pattern = glob.pattern;
+    for (var i = 0; i != pattern.length; ++i) {
+      final char = pattern.codeUnitAt(i);
+      if (_specialGlobCharsCodeUnits.contains(char)) {
```

Comment:
How about a `switch` to avoid `List.contains`?:

```dart
  switch (char) {
    case 42:  // *
    case 63:  // ?
    case 91:  // [
    case 123: // {
    case 92:  // \
      // handle special glob char
      break;
  }
```

Reply:
It searching a constant number of entries both will theoretically be O(1) --- that being said I agree: this field is inside a tight loop, but because it's a get of a final field it will add code that checks if the field has been initialized (again, inside the loop). Additionally it will have to load data to do the comparisons (vs just executing hard-coded comparisons).
Additionally, with the list only being used here it will be easier to see what actually happens.

Reply:
Thanks!

I like that code better, done.

The fact that there's no way to avoid "magic numbers" bugs me though, the best I could come up with was to add a test

```
expect(r'*?[\{'.codeUnits, [42, 63, 91, 92, 123]);
```

and refer to that in the comment.